compiler/src/dotty/tools/dotc/core/tasty/TreeUnpickler.scala

```diff
@@ -1184,15 +1184,39 @@ class TreeUnpickler(reader: TastyReader,
             case SELECTin =>
               var sname = readName()
               val qual = readTerm()
-              val owner = readType()
-              def select(name: Name, denot: Denotation) =
-                val prefix = ctx.typeAssigner.maybeSkolemizePrefix(qual.tpe.widenIfUnstable, name)
-                makeSelect(qual, name, denot.asSeenFrom(prefix))
-              sname match
-                case SignedName(name, sig, target) =>
-                  select(name, owner.decl(name).atSignature(sig, target))
-                case name =>
-                  select(name, owner.decl(name))
+              val ownerTpe = readType()
+              val owner = ownerTpe.typeSymbol
+              val SignedName(name, sig, target) = sname: @unchecked // only methods with params use SELECTin
+              val qualType = qual.tpe.widenIfUnstable
+              val prefix = ctx.typeAssigner.maybeSkolemizePrefix(qualType, name)
+
+              def overloadNotFound(select: Tree, symPart: String) =
+                typer.ErrorReporting.errorTree(select,
+                  i"""While unpickling, I was looking in `$qual` (of widened type: ${qual.tpe.widen})
+                     |but could not find method `$name` (${sname.info}):
+                     | - $symPart.""".stripMargin)
+
+              val denot =
+                val cls = ownerTpe.classSymbol
+                if !cls.exists then NoDenotation
+                else
+                  val clsDenot = cls.asClass.classDenot
+                  val d0 = clsDenot.findMember(name, cls.thisType, EmptyFlags, EmptyFlags).atSignature(sig, target)
```

Comment:
I would keep using `owner.decl` as a fast path in the common situation where the definition has not been moved to a parent class. If that fails we can fallback to `findMember`, and I think it's fine to always exclude Private since a separate compilation unit cannot see private members anyway (qualified private don't get the Private flag).

Reply:
done